crates/aptos-crypto/src/bls12381.rs

```diff
@@ -0,0 +1,381 @@
+            sigs.push(&sig.sig);
+        }
+
+        let result = blst::min_sig::Signature::verify_multiple_aggregate_signatures(
+            &msgs_refs[..],
+            DST,
+            &pubkeys[..],
+            false,
+            &sigs[..],
+            true,
+            &rands,
+            64,
+        );
+        if result == BLST_ERROR::BLST_SUCCESS {
+            Ok(())
+        } else {
+            Err(anyhow!("{:?}", result))
+        }
+    }
+}
+
+impl ValidCryptoMaterial for BLS12381Signature {
+    fn to_bytes(&self) -> Vec<u8> {
+        self.to_bytes().to_vec()
+    }
+}
+impl Length for BLS12381Signature {
+    fn length(&self) -> usize {
+        Self::LENGTH
+    }
+}
+
+impl TryFrom<&[u8]> for BLS12381Signature {
+    type Error = CryptoMaterialError;
+
+    fn try_from(bytes: &[u8]) -> std::result::Result<BLS12381Signature, CryptoMaterialError> {
+        Ok(Self {
+            sig: blst::min_sig::Signature::from_bytes(bytes)
+                .map_err(|_| CryptoMaterialError::DeserializationError)?,
+        })
+    }
+}
+
+impl std::hash::Hash for BLS12381Signature {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        let encoded_signature = self.to_bytes();
+        state.write(&encoded_signature);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        bls12381::*,
+        test_utils::{KeyPair, TestAptosCrypto},
+    };
+    use rand_core::OsRng;
+
+    #[test]
+    fn bls12381_single_message() {
+        let message = b"Hello world";
+        let mut rng = OsRng;
+        let key_pair = KeyPair::<BLS12381PrivateKey, BLS12381PublicKey>::generate(&mut rng);
+        let signature = key_pair.private_key.sign_arbitrary_message(message);
+        assert!(signature
+            .verify_arbitrary_msg(message, &key_pair.public_key)
+            .is_ok());
+    }
+
+    #[test]
+    fn bls12381_batch_verify() {
+        let message = TestAptosCrypto("lalala".into());
+        let mut rng = OsRng;
+        let mut key_pairs = vec![];
+        for _ in 0..1000 {
+            key_pairs.push(KeyPair::<BLS12381PrivateKey, BLS12381PublicKey>::generate(
+                &mut rng,
+            ));
+        }
+        let mut keys_and_signatures = vec![];
+        for keys in key_pairs {
+            let signature = keys.private_key.sign(&message);
+            keys_and_signatures.push((keys.public_key.clone(), signature));
+        }
+        assert!(BLS12381Signature::batch_verify(&message, keys_and_signatures).is_ok());
+    }
+
+    #[test]
+    fn bls12381_aggregate_verify() {
+        let mut rng = OsRng;
+        let message = TestAptosCrypto("lalala".into());
+        let mut key_pairs = vec![];
+        for _ in 0..1000 {
+            key_pairs.push(KeyPair::<BLS12381PrivateKey, BLS12381PublicKey>::generate(
+                &mut rng,
+            ));
+        }
+        let mut signatures = vec![];
+        let mut pubkeys = vec![];
+        for keys in key_pairs.iter().step_by(2) {
```

Comment:
You are skipping over some of these to simulate a n/2 out of n multisig, I assume?